gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala

```diff
@@ -242,28 +243,21 @@ case class WholeStageTransformer(child: SparkPlan, materializeInput: Boolean = f
        * rather than genFinalStageIterator will be invoked
        */
 
-      // If these are two scan transformers, they must have same partitions,
-      // otherwise, exchange will be inserted.
-      val allScanPartitions = basicScanExecTransformers.map(_.getPartitions)
-      val allScanPartitionSchemas = basicScanExecTransformers.map(_.getPartitionSchemas)
-      val partitionLength = allScanPartitions.head.size
-      if (allScanPartitions.exists(_.size != partitionLength)) {
-        throw new GlutenException(
-          "The partition length of all the scan transformer are not the same.")
-      }
+      val allScanSplitInfos = getSplitInfosFromScanTransformer(basicScanExecTransformers)
       val (wsCxt, substraitPlanPartitions) = GlutenTimeMetric.withMillisTime {
         val wsCxt = doWholeStageTransform()
 
-        // the file format for each scan exec
-        val fileFormats = basicScanExecTransformers.map(ConverterUtils.getFileFormat)
-
         // generate each partition of all scan exec
-        val substraitPlanPartitions = (0 until partitionLength).map(
-          i => {
-            val currentPartitions = allScanPartitions.map(_(i))
-            BackendsApiManager.getIteratorApiInstance
-              .genFilePartition(i, currentPartitions, allScanPartitionSchemas, fileFormats, wsCxt)
-          })
+        val substraitPlanPartitions = allScanSplitInfos.zipWithIndex.map {
+          case (splitInfos, index) =>
+            wsCxt.substraitContext.initSplitInfosIndex(0)
+            wsCxt.substraitContext.setSplitInfos(splitInfos)
+            val substraitPlan = wsCxt.root.toProtobuf
+            GlutenPartition(
+              index,
+              substraitPlan.toByteArray,
+              splitInfos.head.preferredLocations().asScala.toArray)
```

Comment:
splitInfos(i).preferredLocations ?

Reply:
Currently, the `GlutenPartition` only has one `preferredLocations`. However, within `GlutenPartition`, there are multiple `SplitInfo` instances, and each `SplitInfo` may have a different `preferredLocations`. However, the existing code only takes the first `preferredLocations`.

https://github.com/oap-project/gluten/blob/f21a96f727ab8368d1de0de25101b41c61d58038/backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala#L126

Reply:
I see, it's for one partition but for several scans. I think the currently implementation loses some of scan preferredLocations. Shall we use `splitInfos.flatMap(_.preferredLocations).distinct` ?

Reply:
Yes. Merge all preferredLocations is better. I will fix.

Reply:
Additional note: distinct is not need as Spark will count the locations to determine which locations are preferred.